Consistently wrap errors in load_cert_chain

sfreilich opened the pull request:
This wraps OpenSSL.SSL.Error with ssl.SSLError in
PyOpenSSLContext.load_cert_chain, similar to the error handling in
other methods of PyOpenSSLContext.

sfreilich commented:
Sorry, I thought I'd fixed that typo before, and not sure how it got back into my local checkout before push.

pquentin reviewed (commented):
Thanks! This looks good. Can you please add a changelog entry? If you need that soon, you'll also have to backport to the 1.26.x branch.

nateprewitt commented:
It looks like we're consistently failing to raise an error with the new test on macOS. That may need a quick check on root cause. If we're not concerned, we may skip that test for the darwin platform.

sfreilich commented:
Filed issue #2628. (But I shouldn't just used the PR number? Sorry just a bit tired and getting the instructions mixed up.)

Noticed there's a typo in my change description, too.

sfreilich commented:
@nateprewitt Interesting, is there a more consistent way to get the underlying error on Mac? Otherwise, how should I skip the test on that platform?

nateprewitt commented:
It looks like macOS may not be raising an error at all:

```python
_________________ TestClientCerts.test_load_invalid_cert_file __________________
Traceback (most recent call last):
  File "/Users/runner/work/urllib3/urllib3/test/with_dummyserver/test_socketlevel.py", line 367, in test_load_invalid_cert_file
    certfile=self.password_key_path,
  File "/Users/runner/work/urllib3/urllib3/.nox/test-3-7/lib/python3.7/site-packages/_pytest/python_api.py", line 952, in __exit__
    fail(self.message)
  File "/Users/runner/work/urllib3/urllib3/.nox/test-3-7/lib/python3.7/site-packages/_pytest/outcomes.py", line 1[96](https://github.com/urllib3/urllib3/runs/6821420445?check_suite_focus=true#step:7:97), in fail
    raise Failed(msg=reason, pytrace=pytrace)
Failed: DID NOT RAISE <class 'ssl.SSLError'>
```

We'd need to do some quick testing to confirm if the test doesn't raise an error or if it raises a different error. I'm assuming it's the former but it would be good to check. Once we have that answer, the next question is should `load_cert_chain` be raising an error? I don't immediately know the answer for that without some digging.

If we do find this is expected behavior, we can add a skip similar to [this](https://docs.pytest.org/en/7.1.x/how-to/skipping.html#id1) based on the host platform.

i.e.
```python
@pytest.mark.skipif(sys.platform == 'darwin', reason='No error raised on macOS')
```

sethmlarson commented:
@sfreilich Thanks for creating this. I've asked our team for anyone with macOS to give the test suite a spin to see if they can see the error since this is a SecureTransport test.

sethmlarson commented:
Oh right, our SecureTransport `load_cert_chain` call is lazy and waits until we're doing a handshake. Let's skip the test on macOS if @nateprewitt and @pquentin are okay with this approach?

sethmlarson commented:
@sfreilich Your latest commit skips the test on macOS completely, however we actually only want to skip when SecureTransport is in use as the test should run and pass on macOS w/ Python's ssl and pyOpenSSL. We have the [`notSecureTransport` pytest annotation](https://github.com/urllib3/urllib3/blob/main/test/__init__.py#L158) for this purpose, could you change to using this instead?

sfreilich commented:
Updated my patch. I think skipping the test in Mac sounds reasonable if the underlying library is just deferring the validation that would raise an exception in that spot. The eventual call to cnx.do_handshake() is already covered by similar error-wrapping logic.

sfreilich commented:
@sethmlarson Missed your comment when I wrote my last. Done.

nateprewitt reviewed (dismissed):
Great, thanks @sfreilich! This looks good from my concern.

nateprewitt commented:
Looks like the PR may need another quick rebase after our merges this morning.

sethmlarson commented:
@nateprewitt You can squash this once it's passing CI, but remove my name from the commit please :) Didn't realize GitHub's rebase mode would add the rebase button pusher to every commit.

nateprewitt commented:
Hmm, it's not clear squashing will remove you from the attribution, @sethmlarson. It doesn't have the normal co-author reference in the extended comment. I can merge as is and we hope, or we'll need @sfreilich to rebase themselves.

sfreilich commented:
I tried, but it doesn't seem to have done anything, probably because this branch already has those commits. Not sure what I can do to fix that.

nateprewitt commented:
You can try squashing them yourself and force pushing if you're comfortable doing that. Otherwise, we can just merge at this point.

sfreilich commented:
@nateprewitt Done. Needs reapproval.

nateprewitt reviewed (approved):
Perfect, thanks again @sfreilich!

sethmlarson commented:
Thanks @nateprewitt and @sfreilich for fighting through the mess I made ;)

pquentin commented:
Does someone understand why a change in the pyOpenSSL backend affected the SecureTransport tests? This still baffles me.

sfreilich commented:
The code change doesn't effect the SecureTransport tests. I added a new test that asserts that errors in load_cert_chain are consistently raised. (And also modified an existing test that had been working around the inconsistency to expect consistent behavior, as it should have.) In my first draft, the new test was applied to all the SSL backends. But it failed for SecureTransport because the underlying library defers some validation, so it doesn't raise an exception in the situation the new test asserts about. So the new test was added just for the other backends.

pquentin commented:
Ok I get it now, thanks for the explanation!
